aws/resource_aws_cloudformation_type.go

```diff
@@ -0,0 +1,353 @@
+
+		Schema: map[string]*schema.Schema{
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"default_version_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"deprecated_status": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"documentation_url": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"execution_role_arn": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validateArn,
+			},
+			"is_default_version": {
+				Type:     schema.TypeBool,
+				Computed: true,
+			},
+			"logging_config": {
+				Type:     schema.TypeList,
+				Optional: true,
+				ForceNew: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"log_group_name": {
+							Type:     schema.TypeString,
+							Required: true,
+							ForceNew: true,
+							ValidateFunc: validation.All(
+								validation.StringLenBetween(1, 512),
+								validation.StringMatch(regexp.MustCompile(`[\.\-_/#A-Za-z0-9]+`), "must contain only alphanumeric, period, hyphen, forward slash, and octothorp characters"),
+							),
+						},
+						"log_role_arn": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ForceNew:     true,
+							ValidateFunc: validateArn,
+						},
+					},
+				},
+			},
+			"provisioning_type": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"schema": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"source_url": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"schema_handler_package": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringMatch(regexp.MustCompile(`^(https|s3)\:\/\/.+`), "must begin with s3:// or https://"),
+			},
+			"type": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringInSlice(cloudformation.RegistryType_Values(), false),
+			},
+			"type_arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"type_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.All(
+					validation.StringLenBetween(10, 204),
+					validation.StringMatch(regexp.MustCompile(`[A-Za-z0-9]{2,64}::[A-Za-z0-9]{2,64}::[A-Za-z0-9]{2,64}(::MODULE){0,1}`), "three alphanumeric character sections separated by double colons (::)"),
+				),
+			},
+			"version_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"visibility": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsCloudFormationTypeCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*AWSClient).cfconn
+
```

Comment:
Suggest making local variable `typeName` (as it's a required attribute) and using in error messages.

```suggestion
	typeName := d.Get("type_name").(string)
```

Reply:
We used to do this more, but the APIs can change over time, so then you wind up needing the other style anyways. Adjusted it though.